Repair - Improve location of repair interaction points

Ivanowicz opened the pull request:
**When merged this pull request will:**
- Fix #9558 
- Add names of hitpoints with `depends` parameter to the name of the hitpoint they depend on
- Expand ignored hipoint list to include armor parts, sideskirts and missing slats
- 
### IMPORTANT

- If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

Ivanowicz commented:
The only helicopter I'm still having issues with is the RHS AH-1Z, the hull and tail hitpoints are missing. Those hitpoints have empty selection strings, but mi-8 config is also missisng them and the hitpoints do show up, so I'm currently out of ideas.

LinkIsGrim commented on addons/repair/functions/fnc_addRepairActions.sqf:
```diff
@@ -119,6 +119,10 @@ private _turretPaths = ((fullCrew [_vehicle, "gunner", true]) + (fullCrew [_vehi
 
         // Find the action position
         private _position = compile format ["_target selectionPosition ['%1', 'HitPoints'];", _selection];
+        if ("rotor" in _hitpoint) then {
```
is this naming convention enforced by the engine or is GM gonna be special again?

Ivanowicz replied:
oh, haven't even thought of that - sadly I have no idea and no way to check GM
all the testing I've done was on vanilla and RHS, there it seems that 100% rotors do have rotor somewhere in the name.

Ivanowicz replied:
Seems to be enforced by the engine https://community.bistudio.com/wiki/Arma_3:_Hitpoints

Ivanowicz replied:
Nevermind, I think I misread something - running the script from the link shows that just with RHS the hitpoint list grows to 600+, so this will have to be checked for GM (RHS and CUP seem ok here)

LinkIsGrim commented:
> The only helicopter I'm still having issues with is the RHS AH-1Z, the hull and tail hitpoints are missing. Those hitpoints have empty selection strings, but mi-8 config is also missisng them and the hitpoints do show up, so I'm currently out of ideas.
> 

IIRC they use armorComponent

Ivanowicz commented:
> > The only helicopter I'm still having issues with is the RHS AH-1Z, the hull and tail hitpoints are missing. Those hitpoints have empty selection strings, but mi-8 config is also missisng them and the hitpoints do show up, so I'm currently out of ideas.
> 
> IIRC they use armorComponent

Yup, exactly, that's why it's so weird that Mi-8 works. My best guess is that there's something weird in the config that causes it to skip over the check in `fnc_getSelectionsToIgnore.sqf` at line 100, but I just can't figure out what.

Ivanowicz commented:
Added `hull` and `engine` to hitpoints using `AveragePoint` - I feel that those seem to work better in most cases. This can't really be used for all hitpoints, since it causes way more overlapping interaction points than the old method.

Fixed an error related to full repair of vehicles that return an empty `getAllHitPointsDamage` array - there are a few (less than 10 I think) vehicles like that in CUP. Should we handle them in some way to allow full repair?

Also finally figured out what was the issue with AH-1Z - there are some hitpoints which have `depends` specified, but the specified hitpoint is ignored (no selection or armorComponent), causing the hitpoint with `depends` to be unrepairable. In this case it's the tail rotor which **can** be damaged, but it `depends` on tail, which is ignored.

rautamiekka commented:
> empty `getAllHitPointsDamage` array - there are a few (less than 10 I think) vehicles like that in CUP. Should we handle them in some way to allow full repair?

Another item to the TODO list to fix in CUP, which they must do.

rautamiekka commented on addons/repair/functions/fnc_addRepairActions.sqf:
```diff
@@ -31,7 +31,7 @@ if (_type in _initializedClasses) exitWith {};
 
 // get selections to ignore
-private _selectionsToIgnore = _vehicle call FUNC(getSelectionsToIgnore);
+([_vehicle] call FUNC(getSelectionsToIgnore)) params["_selectionsToIgnore", "_dependsArray"];
```
```suggestion
([_vehicle] call FUNC(getSelectionsToIgnore)) params ["_selectionsToIgnore", "_dependsArray"];
```

rautamiekka commented on addons/repair/functions/fnc_addRepairActions.sqf:
```diff
@@ -119,6 +119,10 @@ private _turretPaths = ((fullCrew [_vehicle, "gunner", true]) + (fullCrew [_vehi
 
         // Find the action position
         private _position = compile format ["_target selectionPosition ['%1', 'HitPoints'];", _selection];
+        if ("rotor" in _hitpoint || "hull" in _hitpoint || "engine" in _hitpoint) then {
```
I wonder if RegEx is faster ? I think it's worth testing. Like this:
```suggestion
        if (_hitpoint regexMatch ".*?(?:rotor|hull|engine).*+/o") then {
```

Ivanowicz replied:
Regex is 2x faster here, but the performance is in one-thousands of miliseconds, so leaving it as is is probably fine.

rautamiekka commented on addons/repair/functions/fnc_canFullRepair.sqf:
```diff
@@ -0,0 +1,33 @@
+TRACE_2("params",_unit,_target);
+
+private _return = false;
+
+if ((getAllHitPointsDamage _target) isNotEqualTo []) then {
+	if (((getAllHitPointsDamage _target) select 2) findIf {_x > 0} != -1) then {
```
```suggestion
private _allHitPointsDamage = getAllHitPointsDamage _target;

if (_allHitPointsDamage isNotEqualTo []) then {
	if ((_allHitPointsDamage select 2) findIf {_x > 0} != -1) then {
```

rautamiekka commented on addons/repair/functions/fnc_getSelectionsToIgnore.sqf:
```diff
@@ -128,6 +136,9 @@ private _processedSelections = [];
             ERROR_1("group: %1",_hitpointGroups # _groupIndex);
         };
+        private _parentHitpoint = getText (_vehCfg >> "HitPoints" >> _hitpoint >> "depends");
+        _parentHitpoint = toLower _parentHitpoint;
```
```suggestion
        _parentHitpoint = toLowerANSI _parentHitpoint;
```

LinkIsGrim replied:
why?

LinkIsGrim commented on addons/repair/functions/fnc_getSelectionsToIgnore.sqf:
```diff
@@ -20,7 +20,12 @@ params ["_vehicle"];
 private _initializedClasses = missionNamespace getVariable [QGVAR(hitPointsToIgnoreInitializedClasses), createHashMap];
-if (_type in _initializedClasses) exitWith {_initializedClasses get _type};
+private _initializedDepends = missionNamespace getVariable [QGVAR(dependsHitPointsInitializedClasses), createHashMap];
+if (_type in _initializedClasses) exitWith 
+{
```
```suggestion
if (_type in _initializedClasses) exitWith {
```

LinkIsGrim commented on addons/repair/functions/fnc_getSelectionsToIgnore.sqf:
```diff
@@ -20,7 +20,11 @@ params ["_vehicle"];
 TRACE_2("getSelectionsToIgnore",_vehicle,_type);
 private _initializedClasses = missionNamespace getVariable [QGVAR(hitPointsToIgnoreInitializedClasses), createHashMap];
-if (_type in _initializedClasses) exitWith {_initializedClasses get _type};
+private _initializedDepends = missionNamespace getVariable [QGVAR(dependsHitPointsInitializedClasses), createHashMap];
```
why is this separate?

Ivanowicz commented:
The latest commits should solve the problems from the 2 reviews above. Since now depends are properly handled, they don't have to be always added to ignored hitpoints, which also means there's no need to make those multi-hitpoint names.

I noticed some RHS tanks have smokelauncher hitpoints - should they be left repairable?

rautamiekka commented:
> I noticed some RHS tanks have smokelauncher hitpoints - should they be left repairable?

Why not ? TBF I could see rearming doing it simultaneously and it really being useful while rearming (depending on whether rearming is only possible with a rearm vehicle), but it shouldn't hurt having the option regardless. At least it gives modders another way to expand, if anything.

rautamiekka commented on addons/repair/functions/fnc_getSelectionsToIgnore.sqf:
```diff
@@ -63,8 +67,11 @@ private _processedSelections = [];
 
-    if (_hitpoint select [0,7] isEqualTo "hitera_" || {_hitpoint select [0,8] isEqualTo "hitslat_"} || {_hitpoint select [0,4] isEqualTo "era_"}) then { // skip era/slat
-        TRACE_3("Skipping ERA/Slat HitPoint",_hitpoint,_forEachIndex,_selection);
+    if (_hitpoint select [0,7] isEqualTo "hitera_" || {_hitpoint select [0,4] isEqualTo "era_"}
+    || {_hitpoint select [0,8] isEqualTo "hitslat_"}  || {_hitpoint select [0,5] isEqualTo "slat_"}
+    || {_hitpoint select [0,9] isEqualTo "sideskirt"} || {_hitpoint select [0,6] isEqualTo "armor_"}
+    || {_hitpoint select [0,3] isEqualTo "mud"}) then { // skip era/slat/sideskirt/armor/mudguards
```
Hoo boy, I imagine the performance hit is pretty high here ... What about RegEx ?
```suggestion
    if (_hitpoint regexMatch "^(?:(?:(?:hit)?(?:era|slat)|armor)_|sideskirt|mud).*+/o") then { // skip era/slat/sideskirt/armor/mudguards
```

Ivanowicz replied:
4x faster with RegEx (0.0004 ms)
Just to make sure - is this the correct way to check performance?
`["_hitpoint regexMatch '^(?:(?:(?:hit)?(?:era|slat)|armor)_|sideskirt|mud).*+/o'",[{_hitpoint = "hitpoint_test"}],10000] call BIS_fnc_codePerformance;`

Ivanowicz commented:
> Why not ?

I noticed some of them had complex `depends`, so I wanted to make sure if it was worth digging into it.
And now that I look closer, not only do most of them have multiple parents but they also have `depends` in a chain, e.g.
smoketube_1_hitpoint `depends = "era_22_hitpoint + smoketube_2_hitpoint";`
smoketube_2_hitpoint `depends = "era_22_hitpoint + smoketube_3_hitpoint";`

Not sure how to handle that to be honest.

rautamiekka commented:
> > Why not ?
> 
> I noticed some of them had complex `depends`, so I wanted to make sure if it was worth digging into it. And now that I look closer, not only do most of them have multiple parents but they also have `depends` in a chain, e.g. smoketube_1_hitpoint `depends = "era_22_hitpoint + smoketube_2_hitpoint";` smoketube_2_hitpoint `depends = "era_22_hitpoint + smoketube_3_hitpoint";`
> 
> Not sure how to handle that to be honest.

Ahhh, fair enough, no wonder, then.

rautamiekka commented on addons/repair/functions/fnc_setHitPointDamage.sqf:
```diff
@@ -26,40 +26,36 @@ TRACE_4("params",_vehicle,typeOf _vehicle,_hitPointIndex,_hitPointDamage);
-            if (_forEachIndex == _hitPointIndex) then {
-                _hitPointDamageRepaired = (_allHitPointDamages select _forEachIndex) - _hitPointDamage;
-            };
+    if (!(_forEachIndex in _selectionsToIgnore) && isNil{_dependsIndexMap get _forEachIndex}) then {
```
```suggestion
    if (!(_forEachIndex in _selectionsToIgnore) && !(_forEachIndex in _dependsIndexMap)) then {
```

rautamiekka commented on addons/repair/functions/fnc_setHitPointDamage.sqf:
```diff
@@ -26,40 +26,36 @@ TRACE_4("params",_vehicle,typeOf _vehicle,_hitPointIndex,_hitPointDamage);
 
 // exit if the hitpoint is not valid
-if ((_hitPointIndex < 0) || {_hitPointIndex >= (count _allHitPoints)}) exitWith {ERROR_2("NOT A VALID HITPOINT: %1-%2", _hitPointIndex,_vehicle);};
+if ((_hitPointIndex < 0) || {_hitPointIndex >= (count _hitSelections)}) exitWith {ERROR_2("NOT A VALID HITPOINT: %1-%2", _hitPointIndex,_vehicle);};
```
```suggestion
if ((_hitPointIndex < 0) || {_hitPointIndex >= (count _hitSelections)}) exitWith {ERROR_2("NOT A VALID HITPOINT: %1-%2",_hitPointIndex,_vehicle);};
```

rautamiekka commented on addons/repair/functions/fnc_normalizeHitPoints.sqf:
```diff
@@ -21,47 +21,11 @@ TRACE_2("params",_vehicle, typeOf _vehicle);
-    missionNamespace setVariable _x;
-} forEach _savedGlobals;
+    private _damage = _vehicle getHitIndex _y;
+    TRACE_2("setting depends hitpoint", _x, _damage);
```
```suggestion
    TRACE_2("setting depends hitpoint",_x,_damage);
```

rautamiekka commented on addons/repair/functions/fnc_getSelectionsToIgnore.sqf:
```diff
@@ -147,7 +160,7 @@ private _processedSelections = [];
-_initializedClasses set [_type, _indexesToIgnore];
-missionNamespace setVariable [QGVAR(hitPointsToIgnoreInitializedClasses), _initializedClasses];
+_initializedClasses set [_type, [_indexesToIgnore, _dependsIndexMap]];
+SETMVAR(GVAR(ignoredAndDependsInitializedClasses), _initializedClasses);
```
```suggestion
SETMVAR(GVAR(ignoredAndDependsInitializedClasses),_initializedClasses);
```

rautamiekka commented on addons/repair/functions/fnc_getSelectionsToIgnore.sqf:
```diff
@@ -19,8 +19,11 @@ params ["_vehicle"];
 TRACE_2("getSelectionsToIgnore",_vehicle,_type);
-private _initializedClasses = missionNamespace getVariable [QGVAR(hitPointsToIgnoreInitializedClasses), createHashMap];
-if (_type in _initializedClasses) exitWith {_initializedClasses get _type};
+private _initializedClasses = GETMVAR(GVAR(ignoredAndDependsInitializedClasses), createHashMap);
```
```suggestion
private _initializedClasses = GETMVAR(GVAR(ignoredAndDependsInitializedClasses),createHashMap);
```

rautamiekka commented on addons/repair/functions/fnc_doRepair.sqf:
```diff
@@ -54,11 +70,16 @@ if (isArray _hitpointGroupConfig) then {
+                        private _subPointCurDamage = _vehicle getHitIndex _hitPointIndex;
+                        private _subPointNewDamage = (_subPointCurDamage - 0.5) max _postRepairDamageMin;
+                        if (_subPointNewDamage < _subPointCurDamage) then {
+                            TRACE_3("repairing sub point", _vehicle, _subHitIndex, _subPointNewDamage);
```
```suggestion
                            TRACE_3("repairing sub point",_vehicle,_subHitIndex,_subPointNewDamage);
```

rautamiekka commented on addons/repair/functions/fnc_doRepair.sqf:
```diff
@@ -34,10 +36,24 @@ private _hitPointCurDamage = _vehicle getHitIndex _hitPointIndex;
+            _parentHitpointNewDamage = (_parentHitpointCurDamage - 0.5) max _postRepairDamageMin;
+        };
+        _repairedHitpoints pushBack _parentIndex;
+        TRACE_4("Repairing depends and parent", _vehicle, _hitPointIndex, _parentIndex, _parentHitpointNewDamage);
```
```suggestion
        TRACE_4("Repairing depends and parent",_vehicle,_hitPointIndex,_parentIndex,_parentHitpointNewDamage);
```

Ivanowicz commented:
Added handling of depends with complex parents (e.g. HitEngine `depends = 0.5 * (HitEngine1 + HitEngine2)`).
Also did some cleanup (mostly changing `hitpoint` to `hitPoint` for consistency), I hope it's not too out of scope for this pull request.

I tested RHS, CUP and vanilla, but I don't have any of the CDLCs so those still need checking.
Here's a script for saving unique depends - parent pairs, so finding any weird ones should be quick.
https://pastebin.com/Q6yMz7Cv

rautamiekka commented on addons/repair/functions/fnc_canRepair.sqf:
```diff
@@ -35,10 +35,10 @@ private _engineerRequired = if (isNumber (_config >> "requiredEngineer")) then {
 
 private _items = _config call FUNC(getRepairItems);
-if (count _items > 0 && {!([_caller, _items] call FUNC(hasItems))}) exitWith {false};
+if (count _items > 0 && {!([_unit, _items] call FUNC(hasItems))}) exitWith {false};
```
I think it'd be much simpler to have `hasItems` do `count _items` instead, less remembering stuff like this needed that way.

johnb432 commented:
I'm going to revert a number of variable name changes, as they make the PR harder to review.

johnb432 commented on addons/repair/functions/fnc_canFullRepair.sqf:
This should be in a separate PR.
